apps/src/blockly/addons/cdoSerializationHelpers.js

```diff
@@ -0,0 +1,184 @@
+) {
+  if (!workspace.rendered) {
+    return;
+  }
+  const {contentWidth = 0, viewWidth = 0} = workspace.getMetrics();
+  const padding = viewWidth ? WORKSPACE_PADDING : 0;
+  const width = viewWidth || contentWidth;
+
+  // The "cursor" object tracks a position on the workspace that starts in the top
+  // corner, then moves below each block that gets manually repositioned.
+  let cursor = {
+    x: padding,
+    y: padding,
+  };
+  // If the workspace is RTL, horizontally mirror the starting position.
+  cursor.x = workspace.RTL ? width - cursor.x : cursor.x;
+
+  blocks.forEach(block => {
+    positionBlock(block, cursor);
+  });
+}
+
+/**
+ * Use a cursor to position a block on a workspace (if it does not already have a position)
+ * @param {Blockly.Block} block - the block to be moved
+ * @param {object} cursor - a location for moving a block, if needed
+ * @param {number} cursor.x - an x-coordinate for moving a block
+ * @param {number} cursor.y - a y-coordinate for moving a block
+ * @return {object} the cursor with updated coordinates
+ */
+export function positionBlockWithCursor(block, cursor) {
+  addUnusedFrame(block);
+  if (blockLocationUnset(block)) {
+    block.moveTo(getNewLocation(block, cursor));
+    cursor.y += getCursorYAdjustment(block);
+  }
+  return cursor;
+}
+
+/**
+ * Determines where the current block should be positioned, based on the cursor
+ * @param {Blockly.Block} block - the block to be moved
+ * @param {object} cursor - a location for moving a block, if needed
+ * @param {number} cursor.x - an x-coordinate for moving a block
+ * @param {number} cursor.y - a y-coordinate for moving a block
+ */
+function getNewLocation(block, cursor) {
+  const blockHasFrameSvg = !!block.functionalSvg_ || !!block.unusedSvg_;
+  const blockTopPadding = blockHasFrameSvg ? SVG_FRAME_TOP_PADDING : 0;
+  const blockSidePadding = blockHasFrameSvg ? SVG_FRAME_SIDE_PADDING : 0;
+  const isRTL = block.workspace.RTL;
+  const newLocation = {
+    x: cursor.x + (isRTL ? -blockSidePadding : blockSidePadding),
+    y: cursor.y + blockTopPadding,
+  };
+  return newLocation;
+}
+
+/**
+ * Determines the amount to move the cursor down, based on the block that was just positioned.
+ * @param {Blockly.Block} block - the block that was just moved
+ * @return {number} - the distance to move the cursor down in preparation for the next move
+ */
+function getCursorYAdjustment(block) {
+  const blockHeight = block.getHeightWidth().height;
+  const blockHasFrameSvg = !!block.functionalSvg_ || !!block.unusedSvg_;
+  const blockVerticalPadding =
+    VERTICAL_SPACE_BETWEEN_BLOCKS + (blockHasFrameSvg ? SVG_FRAME_HEIGHT : 0);
+  return blockHeight + blockVerticalPadding;
+}
+/**
+ * Determines whether a block needs to be repositioned, based on its current position.
+ * @param {Blockly.Block} block - the block being considered
+ * @return {boolean} - true if the block is at the top corner of the workspace
+ */
+function blockLocationUnset(block) {
```

Comment:
The name of this function sounds like it may unset or change the block location instead of checking whether the block needs to be re-positioned - maybe considering something like `isBlockLocationUnset`?